.github/workflows/test-db-changes.yml

```diff
@@ -0,0 +1,62 @@
+name: Test DB Changes
+
+on:
+  push:
+    branches:
+      - main
+      - patch-*
+  pull_request:
+    paths:
+      - 'server/datastore/mysql/schema.sql'
+      - 'server/datastore/mysql/migrations/**.go'
+      - '.github/workflows/test-schema-changes.yml'
+  workflow_dispatch: # Manual
+
+permissions:
+  contents: read
+
+jobs:
+  test-schema-changes:
+    runs-on: ubuntu-latest
+    steps:
+    - name: Install Go
+      uses: actions/setup-go@84cbf8094393cdc5fe1fe1671ff2647332956b1a # v2
+      with:
+        go-version: '^1.19.1'
+    - name: Checkout Code
+      uses: actions/checkout@629c2de402a417ea7690ca6ce3f33229e27606a5 # v2
+      with:
+        fetch-depth: 0
+
+    - name: Start Infra Dependencies
+      # Use & to background this
+      run: docker-compose up -d mysql_test &
+
+    - name: Verify test schema changes
+      run: |
+        make dump-test-schema
+        if [[ $(git diff server/datastore/mysql/schema.sql) ]]; then
+          echo "❌ fail: uncommited changes in schema.sql"
+          echo "please run `make dump-test-schema` and commit the changes"
+          exit 1
+        fi
+
+    # TODO: This doesn't cover all scenarios since other PRs might
+    # be merged into `main` after this check has passed.
+    #
+    # We should add a Slack notification or something similar for
+    # when this check fails on `main`.
+    - name: Check migration order
+      run: |
+        all_migrations=($(ls server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
+        new_migrations=($(git diff --name-only origin/${{github.base_ref}} -- server/datastore/mysql/migrations/tables/20*_*.go | sort -r))
```

Comment:
Why will this work in main in the example you mentioned? Would `github.base_ref` be one of the two PRs somehow?

Reply:
yes, after `actions/checkout` does the git checkout, the current HEAD is what they call the "merge commit" i.e. github.base_ref + the PR changes merged.

So when we run `git diff origin/github.base_ref`, the diff lists only the files modified by the PR.

Reply:
So if this runs on main at any random point, it won't work, it has to run when the merge commit hits main... it won't be caught in our nightly builds unless this happens to be the last commit in main at the time. Am I understanding correctly?

I think this is fine, it'll help us catch more cases, but it's not bullet proof. Not sure if anything beats the test that deploys version A, then migrates to version B and runs fleet serve. But that's a hard one to automate.

Reply:
> We can modify the workflow to use the latest released tag when it's run on main instead of github.base_ref, and I think that will fix this particular issue, right?

I don't think I understand.

> Having said that, if we think that actually doing the migration is a net positive (it might even also catch other cases?) I don't think it's very difficult to implement honestly! (this workflow already has a DB instance running!)

It wouldn't be just running the migration, it would be also starting fleet serve to make sure it runs its usual checks.

The way it would work is:

1. Checkout latest tag and build fleet
2. Run migrations
3. Start fleet serve
4. Wait a bit, if fleet exits on its own, FAIL, otherwise keep going
5. Kill fleet serve
6. Checkout main and build fleet
7. Run migrations
8. Start fleet serve
9. Wait a bit, if fleet exits on its own, FAIL, otherwise SUCCESS

Ideally, we would test the past 2 or 3 releases.

Reply:
> I don't think I understand.

I was thinking on something like this pseudo-code:

```
if base_branch == "main":
  comp = get_latest_tag()
else
  comp = github.base_ref

new_migrations = $("git diff --name-only $comp -- ...")
```

so, if the workflow runs in main, we compare the migrations against the latest released fleet version (or last two, etc). This should cover the case you raised if I understand correctly

> It wouldn't be just running the migration, it would be also starting fleet serve to make sure it runs its usual checks.

yeah, I expressed myself poorly, but I had in mind something akin to what you described (thanks btw!)

I was planning on merging this _after_ the release, so we don't have any surprises during the freeze week, I will see how hard is to do the approach you described before then.

Reply:
> I was thinking on something like this pseudo-code:

Ah, yeah, that could work. Worth giving it a try.

Reply:
@chiiph I updated a branch with what we discussed here, would you mind taking another look? 🙏